persistent_connection.go

```diff
@@ -357,6 +357,8 @@ func (pconn *persistentConn) prepareDataConn() (func() (net.Conn, error), error)
 			return nil, ftpError{err: netErr, temporary: isTemporary}
 		}
 
+		dc.SetDeadline(time.Now().Add(pconn.config.DataTimeout))
```

Comment:
This will only catch timeouts starting the transfer, but not timeouts in the middle of the transfer.  You need to reset your deadline after every read/write (which makes it annoying to use io.Copy). Maybe the best approach is to make io.Reader/Writer wrappers that set the deadline before every read/write call automatically. You just have to be careful to implement the WriteTo and ReadFrom as appropriate to keep io.Copy efficient.